### PR TITLE
Display all collection items on all archive levels in collection list tab

### PR DESCRIPTION
**Pull request by siiriylonen**

### CODE REVIEW

File: config/vufind/config.ini [hunk 1]
```diff
@@ -2159,6 +2159,9 @@ HMACkey = mySuperSecretValue
 route[record] = collection
 route[search2record] = search2collection
 
+; On collection list tab, display all collection items also on lower archive levels
+; showFullHierarchyTree = true
```
Comment (demiankatz) [inline config/vufind/config.ini:2163]: Would it be better to move this into the hierarchy configuration in HierarchyDefault.ini, and then access the setting through the hierarchy driver retrieved from the record driver? This feels like a setting that might better belong in that context, and moving it would also eliminate the need to change factories and constructors, since you could just add a new method to [the ConfigurationBased hierarchy driver](https://github.com/vufind-org/vufind/blob/dev/module/VuFind/src/VuFind/Hierarchy/Driver/ConfigurationBased.php) and then access it using `$driver->getHierarchyDriver()->...`.

File: module/VuFind/src/VuFind/RecordTab/CollectionList.php [hunk 2]
```diff
@@ -126,6 +147,12 @@ public function getResults()
     {
         if (null === $this->results) {
             $driver = $this->getRecordDriver();
+            if ($this->showFullHierarchy) {
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/RecordTab/CollectionList.php:150]: If we need to do special processing based on this logic, might it make more sense to do it directly in [initFromRecordDriver](https://github.com/vufind-org/vufind/blob/cdb1273aaea1d449f2f47cfb884458f2bafe2377/module/VuFind/src/VuFind/Search/SolrCollection/Params.php#L68) instead of manipulating things here? It looks like you might possibly just want to change the way collectionID is set in order to achieve the desired effect.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks for the contribution, @siiriylonen. I'm not entirely sure I understand your use case or what you are trying to accomplish, so if you could tell me more about the context, it might help me do a better job of testing (and possibly suggesting alternatives). However, even without fully understanding the goal, I did have a couple of suggestions that might help you simplify and better organize the feature.

I hope this is helpful. Please let me know if you want to discuss anything further!

**Comment by siiriylonen** [comment]
Thank you for the comments, I’ll go through them. As for context, this was because when you go to the collection list tab on the item page and it’s the top level of the collection, it shows/fetches all items in that collection to search from within the tab. But when you go to the same tab on a lower archive level item, it doesn’t show any or just the next sub-level items (depending on list_type), but never all of them.

So in short, what I was trying to do, was to display all the items of the collection in the collection list tab, no matter on which archive-level you’re visiting the item page on, within that collection.

**Comment by demiankatz** [comment]
@siiriylonen, thanks for the clarification. I now understand what you are trying to do, but I'm not sure I understand why you would want this behavior.

The collection system is designed to work like a hierarchical file system. So, for example, imagine you have an archival collection like this:

John Smith Papers
- Letters
  - 1800-1850
    - Letter 1
  - 1850-1870
    - Letter 2
- Receipts
  - 1800-1850
    - Receipt 1
  - 1850-1870
    - Receipt 2

Each folder of this tree becomes a collection, and the contents tab for that collection shows its immediate children. This allows the user to drill down to more specific sections of the overall collection (or they can use the tree view to freely navigate).

If you completely flatten the collection items tab, then every single layer of the tree will show the same contents:

1800-1850
1800-1850
1850-1870
1850-1870
Letters
Letter 1
Letter 2
Receipts
Receipt 1
Receipt 2

...and this is confusing because important context is lost. I don't see any advantage to allowing navigation of the entire tree if you can't view the contents of the individual levels of the tree.

If your content is structured differently and you don't need the whole tree, is it possible that a better solution would be to change the indexing rules rather than change the behavior of the user interface?

Or is it possible that the new setting recently introduced in #3182 would be helpful for your use case? This doesn't entirely flatten the tree, but it changes things so that when you "search within collection," the search goes all the way down into subcollections, instead of searching only the current layer. In this way, viewing still shows the hierarchy in a filtered fashion, but searches have a broader scope.

And I apologize if I'm missing the point of your changes or over-explaining anything -- I just want to be sure we are on the same page as we move forward with this discussion!

**Comment by siiriylonen** [comment]
Ok, this still seems to be something we’d like to try out, even if it does flatten the hierarchy, but we’ll go about it differently, so I will close this pr. Regardless, thank you for your time and insight.

**Comment by demiankatz** [comment]
Thanks for sharing @siiriylonen, even if it didn't quite work out in this form. If your new solution includes elements that you would like to see in the core code, please feel free to open another PR in the future!
